lib/charms/alertmanager_k8s/v0/alertmanager_remote_configuration.py

```diff
@@ -0,0 +1,408 @@
+
+    The `RemoteConfigurationProvider` object can be instantiated as follows in your charm:
+
+    ```
+    from charms.alertmanager_k8s.v0.alertmanager_remote_configuration import
+        RemoteConfigurationProvider,
+    )
+
+    def __init__(self, *args):
+        ...
+        self.remote_configuration_provider = RemoteConfigurationProvider(
+            charm=self,
+            alertmanager_config=some_config_dict,
+        )
+        ...
+    ```
+
+    The `RemoteConfigurationProvider` assumes that, in the `metadata.yaml` of your charm,
+    you declare a required relation as follows:
+
+    ```
+    provides:
+        remote-configuration:  # Relation name
+            interface: alertmanager_remote_configuration  # Relation interface
+    ```
+
+    The `RemoteConfigurationProvider` provides handling of the most relevant charm
+    lifecycle events. On each of the defined Juju events, Alertmanager configuration and templates
+    from a specified file will be pushed to the relation data bag.
+    Inside the relation data bag, Alertmanager configuration will be stored under
+    `alertmanager_configuration` key, while the templates under the `alertmanager_templates` key.
+    Separation of the main configuration and the templates is dictated by the assumption that
+    the default provider of the `alertmanager_remote_configuration` relation will be
+    `alertmanager-k8s` charm, which requires such separation.
+    """
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        alertmanager_config: Optional[dict] = None,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """API that manages a provided `remote-configuration` relation.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            alertmanager_config: Alertmanager configuration dictionary.
+            relation_name: Name of the relation with the `alertmanager_remote_configuration`
+                interface as defined in metadata.yaml. Defaults to `remote-configuration`.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self.alertmanager_config = alertmanager_config
+        self._relation_name = relation_name
+
+        on_relation = self._charm.on[self._relation_name]
+
+        self.framework.observe(on_relation.relation_joined, self._on_relation_joined)
+
+    def _on_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Event handler for RelationJoinedEvent.
+
+        Takes care of pushing Alertmanager configuration to the relation data bag.
+
+        Args:
+            event: Juju event
+        """
+        if not self._charm.unit.is_leader():
+            return
+        if self.alertmanager_config:
+            self.update_relation_data_bag(self.alertmanager_config, event.relation)
+        else:
+            logger.warning("Alertmanager configuration not available. Ignoring...")
```

Comment:
Sounds like fairly simple change, but why would want to push invalid config to Alertmanager if already at this point you know it's invalid? 
I like the idea of blocking the provider charm.
As for the remote config overriding the local config - currently local config will also be used if remote config is invalid (if a pending config is invalid, `amtool` prevents applying it). I guess we don't wanna change that, right?

Reply:
I also don't think we should push an invalid config and potentially kill Alertmanager for something which is known-invalid versus flagging instead.

What I actually wonder is that `alertmanager_config` is `Optional[Dict]`. Under the hypothetical case of some charm which includes this `Provider` and sets the filepath via a config value, every single execution of any hook will `logger.warning` for something which is probably expected (the admin hasn't set the config value yet).

`load_config_file` also isn't used anywhere, so I'm guessing the intended case is to import that into the charm.

What if, instead, the constructor for the provider had an additional arg for the path, also as an optional string. If a truthy `alertmanager_config` is passed, use that. If one isn't passed, but a file is, try to load that in the constructor. If neither is present, do nothing. Maybe `logger.debug`

`load_config_file` could then go onto the class (also not used elsewhere) as `@staticmethod` so a charm could still:
```python
remote_config = RemoteConfigurationProvider.load_config_file(...)
remote_provider = RemoteConfigurationProvider(self, alertmanager_config=remote_config, ...)
```

Or (I like this better):
```python
class RemoteConfigurationProvider(Object):
   ...
   @classmethod
   with_config_file(cls, charm, config_file: Path):
     return cls(charm, load_config_file(config_file))
```

Reply:
Hi @rbarry82,
The first solution is pretty much how it is handled right now, except that the `load_config_file` is not in the Provider class.
Isn't the second option against the convention saying that libs shouldn't be doing any storage related actions?

Reply:
Considering that the lib isn't actually doing anything with the container, I would say no, but I would also say that the distinction between option #1 and option #2 is completely arbitrary from that POV, since there's no fundamental difference between "this method from a lib loads a file, and we pass that object into a lib" and "a factory takes an option which may load from a file". Both are methods from a lib which are loading from a file.

The difference to me is that the second one logically connects the "floating" method from the lib with its solitary use case, and makes it immediately clear to anyone reading either the library or a charm using it exactly what `load_config_file`'s purpose is.

Reply:
Yeah, that makes sense. 
The purpose of keeping `load_config_files` outside the `Provider` class was to support potential custom implementations of the provider charm (like if someone wants to do something with the config data before putting it to the data bag). That could be a starting point. Then again, perhaps we shouldn't even consider such situation... WDYT?
As for having the factory accepting the path, I'll be happy to add it if there's no objections (@sed-i, @simskij). 
There's just one thing that confuses me a bit - you mentioned adding another optional arg (config file path) to the constructor. Why would you want that if you have a factory handling it?

Reply:
Addition the optional arg supports your current usage pattern (non-factory) plus the other case.

My personal take is that we shouldn't consider the other scenario -- it's kind of a Pandora's box. The preference is generally "we support this set of options, and if you need something else, please file a request and we'll try to be accommodating, which also means we'll support it going forward. If it doesn't make sense or would cause a nightmare in the future, feel free to fork and we'll still try to help, but we're not going to 'officially' support this in any way".

Allowing custom implementations of the Provider charm is great, but it also means potentially opening the door to unsanitized input which breaks the consumer, or something built for expediency rather than best practice which is now crucial to operations. Again, not to say that users can't do it or that we shouldn't support it, but it's kind of a safety guardrail design which says "if you go outside the borders, there be dragons".

Reply:
I prefer to not mix default constructor with alternative constructors.
We could have mutually exclusive kwargs (`def __ini__(self, *, from_dict=None, from_file=None, ...)`) or a union (e.g. `config: Union[dict, str, TextIO]`).

But agree with @rbarry82 about [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Reply:
Which brings us back to the part @simskij told me to remove :rofl: 
For now, one thing seems clear - we shouldn't support custom implementations by exposing public methods for loading files.
Before I proceed with refactoring, I'd like the constructor case to be clarified too ;) If we wanna allow using path, my vote (if it matters at all :rofl:) is on having the factory.

Reply:
👍 for the factory from me. It seems more obvious than "hey, pass in a dict!" without being really explicit about what that dict should look like. A good docstring for the factory should hopefully point users in the right direction

Reply:
Factory added.

Reply:
I don't think it's as much about being Python as it is about UX, preventing footguns, and being able to try to sanitize/validate input a little bit @sed-i. 

It was, from my POV, Pythonic. It just wasn't intuitive or user friendly. For charmers who know exactly what they want, they can easily put together something which will fulfill the type. For users who don't quite know where to get started, it presents another choice they need to investigate before either rejecting it or constructing an implementation which may introduce subtle bugs/be harder to support.